File: recursion/leetcode_Letter_Combinations_of_a_Phone_Number.java
/*Letter_Combinations_of_a_Phone_Number

Given a digit string, return all possible letter combinations that the number
could represent.
A mapping of digit to letters (just like on the telephone buttons) is given
below.
Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
Note:
Although the above answer is in lexicographical order, your answer could be in
any order you want.
*/

import java.util.ArrayList;
import java.util.HashMap;

public class leetcode_Letter_Combinations_of_a_Phone_Number {

    public static void main(String[] args) {
        letterCombinations("2345");
    }

    public static ArrayList<String> letterCombinations(String digits) {
        ArrayList<String> all = new ArrayList<String>();
        ArrayList<String> num = new ArrayList<String>();
        HashMap<Integer, String> phoneMap = new HashMap<Integer, String>();
        phoneMap.put(2, "abc");
        phoneMap.put(3, "def");
        phoneMap.put(4, "ghi");
        phoneMap.put(5, "jkl");
        phoneMap.put(6, "mno");
        phoneMap.put(7, "pqrs");
        phoneMap.put(8, "tuv");
        phoneMap.put(9, "wxyz");
        for (int i = 0; i < digits.length(); ++i) {
            int digit = Integer.parseInt(digits.charAt(i) + "");
            num.add(phoneMap.get(digit));
        }
        nextLetter("", num, all);
		System.out.println(all);
        return all; 
    }

    public static void nextLetter(String acc, ArrayList<String> num, ArrayList<String> all) {
        if (num.size() == 0) {
            all.add(acc);
        } else {
            String str = num.remove(num.size() - 1);
            for (int i = 0; i < str.length(); ++i) {
                nextLetter(str.charAt(i) + acc, num, all);
            }
            num.add(str);
        }
    }
}

File: recursion/leetcode_Letter_Combinations_of_a_Phone_Number.cpp
/*
 * Copyright 2015
 *
 * leetcode Letter Combination of A Phone Number
 */


#include <string>
#include <vector>

using std::string;
using std::vector;


void printVector(const vector<string> &v) {
  for (string str : v) {
    printf("%s, ", str.c_str());
  }
  printf("\n");
}


void printCharVector(const vector<char> &v) {
  for (char c : v) {
    printf("%c ", c);
  }
  printf("\n");
}

class Solution {
 private:
  const vector<string> mapping = {
    "", "*", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

 public:
  vector<string> letterCombinations(string digits) {
    vector<string>* collection = new vector<string>();

    if (digits.length() == 0) {
      return collection;
    }

    vector<string> candicates;
    vector<char>* acc = new vector<char>();

    for (char digit : digits) {
      int index = digit - '0';
      candicates.push_back(Solution::mapping[index]);
    }

    dfs(0, acc, candicates, collection);

    delete acc;

    return *collection;
  }

  void dfs(int i, vector<char>* acc, const vector<string>& candicates,
      vector<string>* collection) {
    if (i == candicates.size()) {
      string combination(acc->begin(), acc->end());
      collection->push_back(combination);
      return;
    }

    for (char c : candicates.at(i)) {
      acc->push_back(c);
      dfs(i + 1, acc, candicates, collection);
      acc->pop_back();
    }
  }
};


int main() {
  Solution sol;
  vector<string> res = sol.letterCombinations("23");

  printVector(res);
}
